laia/decoders/ctc_greedy_decoder.py

```diff
@@ -6,23 +6,54 @@
 
 
 class CTCGreedyDecoder:
-    def __call__(self, x: Any, segmentation: bool = False) -> Dict[str, List]:
+    def __call__(
+        self, x: Any, segmentation: bool = False, apply_softmax: bool = True
+    ) -> Dict[str, List]:
         x, xs = transform_batch(x)
         x = x.detach()
+
+        # Apply softmax to have log-probabilities
+        if apply_softmax:
+            x = torch.nn.functional.log_softmax(x, dim=-1)
+
+        # Transform to list where len(list) = batch_size
         x = [x[: xs[i], i, :] for i in range(len(xs))]
         x = [x_n.max(dim=1) for x_n in x]
+
         out = {}
         if segmentation:
-            out["prob"] = [x_n.values.exp() for x_n in x]
+            out["prob-segmentation"] = [x_n.values.exp() for x_n in x]
             out["segmentation"] = [
                 CTCGreedyDecoder.compute_segmentation(x_n.indices.tolist()) for x_n in x
             ]
+
+        probs = [x_n.values.exp() for x_n in x]
         x = [x_n.indices for x_n in x]
+
         # Remove repeated symbols
-        x = [torch.unique_consecutive(x_n) for x_n in x]
+        zero_tensor = torch.tensor([0]).cuda() if x[0].is_cuda else torch.tensor([0])
```

Comment:
The 2 following code blocks are hard to read due to their density. You can simply let the code "breathe" by adding new lines, or making it more explicit (especially the ternary form above that can hide a bit of complexity.

Reply:
I added some new lines and comments. I also simplified the definition of `zero_tensor` by using `device`.